ARROW-10481: [R] Bindings to add, remove, replace Table columns

jonkeane opened the pull request:
This allows for R users to use either `[[` or `$` syntax for removing, adding, or replacing table columns. 

If one supplies an index larger than the number of columns (e.g. `mtcars[[12]] <- vector`), the column is added to the end, the index is cast to a string (here: `"12"`) for the name. This might be better as an error instead of a silent cast for the name.

nealrichardson commented on r/R/table.R:
```diff
@@ -254,6 +257,49 @@ names.Table <- function(x) x$ColumnNames()
+  if (is.null(value)) {
+    if (is.character(i)) {
+      i <- match(i, names(x)) - 1L
+    }
```
I think we need to do the `- 1L` even if i is not character, otherwise we'll be off by one with an integer `i`.

nealrichardson commented on r/R/table.R:
```diff
@@ -254,6 +257,49 @@ names.Table <- function(x) x$ColumnNames()
+  } else {
+    if (!is.character(i)) {
+      # get or create a/the column name
+      if (i <= ncol(x)) {
```
This is a trivial difference but it does save some work because `ncol == function (x) dim(x)[2L]`

```suggestion
      if (i <= x$num_columns) {
```

jonkeane replied:
I went looking for a function call to `num_columns()` and should have looked for an active binding, oops!

nealrichardson commented on r/R/table.R:
```diff
@@ -254,6 +257,49 @@ names.Table <- function(x) x$ColumnNames()
+    if (!is.character(i)) {
+      # get or create a/the column name
+      if (i <= ncol(x)) {
+        i <- names(x)[[i]]
```
```suggestion
        i <- names(x)[i]
```

nealrichardson commented on r/R/table.R:
```diff
@@ -254,6 +257,49 @@ names.Table <- function(x) x$ColumnNames()
+`$<-.Table` <- function(x, i, value) {
+  assert_that(is.string(i))
+  if (i %in% ls(x)) {
+    assign(i, value, x)
```
Can you add a comment explaining why this `if` is here?

nealrichardson commented on r/tests/testthat/test-Table.R:
```diff
@@ -145,7 +145,55 @@ test_that("[, [[, $ for Table", {
+
+  # can add a named column
+  tab[["new"]] <- letters[10:1]
+  expect_vector(tab[["new"]], letters[10:1])
```
Part of the behavior we want to test is the position of the new column, so let's assert the whole data.frame

```suggestion
  expect_data_frame(tab, cbind(tbl, new = letters[10:1]))
```

nealrichardson commented on r/tests/testthat/test-Table.R:
```diff
@@ -145,7 +145,55 @@ test_that("[, [[, $ for Table", {
+
+  tab$int <- 1:10
+  expect_vector(tab$int, 1:10)
+})
```
Other tests we might want to add (and implement as necessary):

* `tab[[NA]] <- something`
* `tab[[NULL]] <- something`
* R input recycling (making sure we're consistent with how [[<-.data.frame works)
* assigning in an Arrow Array or ChunkedArray. `chunked_array()` should handle that but we should confirm
* any others?

nealrichardson reviewed (changes requested):
Thanks! A few notes

jonkeane commented:
I added the extra tests and some code to get that working. 

I also started down the path of `[<-.Table` though there are more decisions that are under that rock (what to do about row manipulation, etc.), so left it be for now. It currently errors in R (though we probably want to write a more helpful message as part of this PR even if we kick that working to another branch/PR).

nealrichardson commented on r/tests/testthat/test-Table.R:
```diff
@@ -179,10 +183,34 @@ test_that("[[<- assignment", {
+  expect_vector(tab$chunked, 1:10)
+
+  # nonsense indexes
+  expect_error(tab[[NA]] <- letters[10:1], "missing value where TRUE/FALSE needed")
```
I'm not sure that's the right error to raise there so maybe there should be better validation here.

Would also be important to test `NA_integer_` and `NA_real_` because those will not get rejected by the R->cpp type mapping.

nealrichardson commented on r/tests/testthat/test-Table.R:
```diff
@@ -179,10 +183,34 @@ test_that("[[<- assignment", {
+
+  # nonsense indexes
+  expect_error(tab[[NA]] <- letters[10:1], "missing value where TRUE/FALSE needed")
+  expect_error(tab[[NULL]] <- letters[10:1], "argument is of length zero")
```
```suggestion
  expect_error(tab[[NULL]] <- letters[10:1])
```

Since that is an error message coming from base R (right?), we shouldn't assert its exact message because it may be translated to local language (i.e. it only will pass if you're in an English locale).

jonkeane replied:
Yeah, I added error checking into the method call for `[[<-` to match the behavior + catch `NA_integer_`. It looks like the NA checking is actually coming from cpp `arrow::r::Input<R_xlen_t>::type`, if I'm reading correctly. I put a TODO note to see about unifying / centralizing that checking

nealrichardson replied:
Oh nice, I've been bitten by that before, so I guess either I was using `int` instead of `R_xlen_t` or `Rcpp` wasn't as helpful. TIL :)

nealrichardson commented on r/R/table.R:
```diff
@@ -261,28 +261,38 @@ names.Table <- function(x) x$ColumnNames()
       }
     }
 
+    # auto-magic recycling
```
1) We should probably only do recycling if `value` is not an ArrowObject (perhaps with a TODO to support Arrow Scalar recycling)
2) https://vctrs.r-lib.org/reference/vec_recycle.html may be a good solution instead of this bespoke code (we already import vctrs so it's available), or else use `base::rep_len` (cf. https://github.com/apache/arrow/blob/master/r/R/array.R#L291)

jonkeane replied:
TIL about `vctrs::vec_recycle` that's really nice.

nealrichardson reviewed (changes requested):
Thanks for pushing this ahead. 

Re: [<- method, I think a followup JIRA is acceptable. The current error message isn't awesome but is clear that it isn't implemented:

```
Error in tab[1, 1] <- 5 : object of type 'environment' is not subsettable
```

The other thing we should make a followup for is for doing the same thing for RecordBatch. _As it turns out_, that will require a little more C++ work because there is AddColumn and RemoveColumn but no SetColumn in record_batch.h.

jonkeane commented:
Ok, I opened https://issues.apache.org/jira/browse/ARROW-10520 and https://issues.apache.org/jira/browse/ARROW-10521 for the two follow ons, added some more NA and type checking, and replaced the value recycling with {{vctrs::vec_recycle}}

nealrichardson commented on r/R/table.R:
```diff
@@ -254,6 +257,68 @@ names.Table <- function(x) x$ColumnNames()
+  } else if (is.na(i)) {
+    # Catch if a NA_character or NA_integer is passed. These are caught elsewhere
+    # in cpp (i.e. _arrow_RecordBatch__column_name)
+    # TODO: figure out if catching in cpp like ^^^ is preferred
```
If we get a useful error message from the cpp layer, great, use that. Let the static(ish) typing do the work.

jonkeane replied:
It's a mixture, for `NA_integer_` and `NA_real_` the error is the previous "missing value where TRUE/FALSE needed" with `NA_character_` though it removes/creates/replaced a column with the name ``NA``. So I would vote to keep the special casing for now and try catching in cpp or with a helper function we can use across `[[` and `[` methods later.

nealrichardson replied:
I suspect that error comes from https://github.com/apache/arrow/pull/8579/files#diff-242f7f20fb4999f3bbe719d6977d2580231b89334326b588d46e9236c2e28fcdR285, i.e. before we hit cpp.

nealrichardson reviewed (commented):
One note on the TODO but otherwise I think we're good enough for now.

Let's hold this for today see if Romain can merge #8256 tomorrow--I've made enough merge conflicts for him on that PR by now. If it's still not ready then, we can merge.

nealrichardson reviewed (approved):
I've rebased; will merge when CI is green
